Add verification phases visualisation to homepage

maribelferreira opened the pull request:
Create components to display verification process steps:
- Implement PhaseColumn for structured step representation
- Add VerificationStep for individual verification details
- Integrate VerificationPhases component into homepage
- Visualize DNSSEC, X.509, and CMS verification chains

maribelferreira committed:
Add verification phases visualization to homepage

Create components to display verification process steps:
- Implement PhaseColumn for structured step representation
- Add VerificationStep for individual verification details
- Integrate VerificationPhases component into homepage
- Visualize DNSSEC, X.509, and CMS verification chains

maribelferreira committed:
Add tick icon and enable result verification status display

Enhance PhaseColumn component:
- Import and add new tick SVG icon
- Introduce optional `showTick` prop to conditionally render verification status
- Update VerificationPhases to display tick icons for completed verification steps

maribelferreira committed:
Improve VerificationPhases layout with responsive directional arrows

Enhance verification phases visualization:
- Add responsive DirectionalArrow component with vertical/horizontal SVG icons
- Refactor layout to use flex and responsive width for better mobile/desktop display
- Maintain existing verification step details and styling

maribelferreira committed:
Integrate VerificationPhases into HowItWorks section

Refactor homepage layout:
- Move VerificationPhases component into HowItWorks section
- Remove redundant VerificationPhases from index page
- Adjust component styling for better visual integration
- Simplify page structure and remove duplicate content

maribelferreira committed:
Refine DNSSEC verification step details and component props

Update VerificationPhases and related components:
- Make description prop optional in PhaseColumn and VerificationStep
- Adjust DNSSEC step titles and descriptions for clarity
- Hide description on mobile devices in VerificationStep
- Correct domain name from 'caltech.com' to 'caltech.edu'

maribelferreira committed:
Refine verification phases UI spacing and padding

Adjust component styling for improved visual hierarchy:
- Reduce title margin in PhaseColumn
- Modify vertical padding in VerificationPhases
- Decrease padding in VerificationStep for tighter layout

maribelferreira committed:
Enhance responsive typography and layout for verification phases

Adjust component styling for better mobile and desktop experience:
- Reduce text sizes in PhaseColumn and VerificationStep
- Swap title and description order in X.509 verification steps
- Add responsive padding and text scaling across components

maribelferreira committed:
Fine-tune verification phases component styling

Refine responsive design details:
- Adjust text size in PhaseColumn using Tailwind's text sizing
- Remove unused tick icon import
- Reduce padding in VerificationStep
- Simplify DirectionalArrow container padding

maribelferreira committed:
Restructure verification phases component location and imports

Move VerificationPhases to VerificationChain directory:
- Relocate component from Homepage to VerificationChain
- Update import paths in HowItWorks component
- Remove redundant VerificationPhases.astro wrapper
- Add client:load directive to ensure React component hydration

maribelferreira committed:
Add animated verification step border and verification flow

Implement dynamic verification process with:
- Animated gradient border for verifying steps
- Progressive verification state tracking
- Tailwind config for border animation keyframes
- Global CSS for border angle property
- Enhanced PhaseColumn and VerificationStep components to support verification states

maribelferreira committed:
Configure scheme and add @tailwindcss/typography plugin for enhanced text styling

maribelferreira committed:
Add green border for completed verification steps

Enhance PhaseColumn component to visually indicate completed verification phases:
- Conditionally apply green border when step is completed and tick is shown
- Use Tailwind's border utility classes for responsive border styling

maribelferreira committed:
Adjust border width for completed verification steps

Refine PhaseColumn border styling:
- Increase border width from 1 to 2 for default and lg breakpoints
- Maintain green border color for completed steps with tick

maribelferreira committed:
Create reusable Arrow component for responsive directional icons

Implement flexible Arrow component:
- Support multiple arrow directions (up, down, left, right)
- Add responsive direction configuration
- Replace inline SVG arrows in VerificationPhases and VerificationStep
- Use Tailwind classes for responsive visibility

gnarea commented on src/components/VerificationChain/Arrow.tsx:
```diff
@@ -0,0 +1,110 @@
+  direction?: "up" | "down" | "left" | "right";
+  responsiveDirection?: {
+    mobile?: "up" | "down" | "left" | "right";
+    desktop?: "up" | "down" | "left" | "right";
```
There's a lot of repetition here and the risk that these things could get out of sync, so I suggest you create a type alias like `type Direction = "down" | "right";`. Note that I don't see how/where "up" and "left" would be used.

gnarea commented on src/components/VerificationChain/Arrow.tsx:
```diff
@@ -0,0 +1,110 @@
+        className={className}
+      >
+        {getPathsForDirection(direction)}
+      </svg>
```
That SVG is almost entirely triplicated. Please move to a component and parametise what needs to be variable.

gnarea commented on src/components/VerificationChain/Arrow.tsx:
```diff
@@ -0,0 +1,110 @@
+  responsiveDirection,
+  className = "",
+}) => {
+  const getPathsForDirection = (dir: "up" | "down" | "left" | "right") => {
```
```suggestion
  const getPathsForDirection = (dir: Direction) => {
```

... once you add the type alias.

gnarea commented on src/components/VerificationChain/PhaseColumn.tsx:
I'd just call it "Phase". "Column" doesn't add anything, and it's inaccurate when the diagram is rendered vertically on mobile (portrait).

gnarea commented on src/components/VerificationChain/PhaseColumn.tsx:
```diff
@@ -0,0 +1,78 @@
+              d="M17.2584 7.15411C17.0948 6.96724 16.8029 6.94388 16.616 7.11907C15.8919 7.74976 15.2261 8.41549 14.5604 9.10458C13.9064 9.79367 13.264 10.4944 12.645 11.2186C11.734 12.3047 10.858 13.4376 10.0405 14.6056L7.07389 11.4171C6.93374 11.2653 6.70015 11.2186 6.51328 11.3354C6.29137 11.4638 6.22129 11.7441 6.34976 11.966C6.34976 11.966 9.17619 16.8364 9.29298 17.0466C9.40978 17.2451 9.66 ...
+              fill="#191B1A"
+            />
+          </svg>
```
Can this SVG be a separate file?

gnarea commented on src/components/VerificationChain/VerificationPhases.tsx:
```diff
@@ -0,0 +1,206 @@
+  const cmsSteps: VerificationStepData[] = [
+    {
+      title: "Bazinga!",
+      description: "",
```
```suggestion
```

gnarea commented on src/components/VerificationChain/PhaseColumn.tsx:
```diff
@@ -0,0 +1,78 @@
+  isActive?: boolean;
+  verifyingIndex?: number;
+  verifiedIndices?: number[];
+  isCompleted?: boolean;
```
I think this is a complicated and brittle way to represent the current _status_ of the phase. I suggest that you introduce an enum to represent the status, and use it here. Example:

```typescript
enum VerificationStatus {
  PENDING,
  VERIFYING,
  VERIFIED,
}

interface PhaseProps {
  title: string;
  steps: VerificationStepData[];
  result: string;
  status: VerificationStatus;
}
```

With this approach, it becomes impossible to end up with a corrupt state. For example, in the original code, it'd be possible to have `showTick=true` and `isCompleted=false`, which would be a bug. Such corrupt states are now impossible with the enum above.

On the other hand, instead of capturing `verifyingIndex` and `verifiedIndices`, I think it'd be simpler and sufficient to capture the count of verified steps, `verifiedSteps: number`. So the final version of the props would be:

```typescript
interface PhaseProps {
  title: string;
  steps: VerificationStepData[];
  result: string;
  status: VerificationStatus;
  verifiedSteps: number;
}
```

gnarea commented on src/components/VerificationChain/VerificationStep.tsx:
```diff
@@ -0,0 +1,54 @@
+  description,
+  showArrow = false,
+  isVerifying = false,
+  isVerified = false,
```
I'd suggest using the same `VerificationStatus` enum defined elsewhere, so you'll probably have to declare it in its own file `VerificationStatus.ts`.

```suggestion
  status: VerificationStatus;
}

export const VerificationStep: React.FC<VerificationStepProps> = ({
  title,
  description,
  showArrow = false,
  status,
```

gnarea commented on src/components/VerificationChain/VerificationPhases.tsx:
I think most, if not all, comments in this file are redundant because they paraphrase the code they precede. Comments should add context that the reader can't gather from just looking at the code.

gnarea commented on src/components/VerificationChain/VerificationPhases.tsx:
```diff
@@ -0,0 +1,206 @@
+              return newIndices;
+            });
+
+            // Mark this phase as completed after a small delay
+            setTimeout(() => {
+              setCompletedPhases((prev) => {
+                const newCompleted = [...prev];
+                newCompleted[phase] = true;
+                return newCompleted;
+              });
+
+              // If there are more phases, start the next one
+              if (phase < allSteps.length - 1) {
+                setCurrentPhase(phase + 1);
+
+                // Schedule verification of the first step in the next phase
+                setTimeout(() => {
+                  setVerifyingIndices((prev) => {
+                    const newIndices = [...prev];
+                    newIndices[phase + 1] = 0; // Start next phase
+                    return newIndices;
+                  });
+
+                  verifyStep(phase + 1, 0);
+                }, 1000); // Small delay before starting the next phase
+              }
+            }, 500); // Small delay after the last step is verified
+          }
+        }, 2000); // Wait for the animation to complete before marking as verified
+      };
+
+      // Start the verification process
+      verifyStep(0, 0);
+    }, 500); // Initial delay before starting verification
+
+    // Cleanup function
+    return () => {
+      clearTimeout(initialDelay);
+    };
+  }, []); // Empty dependency array means this runs once on mount
```
I'm concerned about the extensive use of `setTimeout()` here, which is problematic. I had a chat with Sonnet 3.7, and it summarised the consequences and alternatives more eloquently than I could, so I'm pasting its response here:

---

# Feedback on Verification Animation Implementation

I've reviewed the verification animation component, and while it effectively demonstrates the VeraId verification process, I believe we can make it more robust and maintainable. Here's some constructive feedback on the current implementation:

## Current Approach: Challenges with setTimeout Chains

The current implementation relies heavily on nested `setTimeout` calls to sequence verification steps. This approach:

1. Creates tightly coupled timing dependencies where changing one timeout requires adjusting multiple others
2. Makes the code difficult to maintain as new steps would require careful timing calculations
3. Creates deeply nested code blocks that reduce readability
4. Makes it hard to pause, resume, or adjust animation speed dynamically
5. Could potentially lead to timing issues on slower devices or when browser tabs are inactive

## Alternative 1: State-Based Approach with useReducer

A more maintainable approach would leverage React's state management to drive the animation sequence:

```jsx
function VerificationPhases() {
  // Define all possible verification steps
  const [state, dispatch] = useReducer(verificationReducer, {
    currentPhase: 0,
    currentStep: 0,
    isAnimating: false,
    stepDelay: 800,
    // other state...
  });
  
  // Effect to handle the animation timing
  useEffect(() => {
    if (!state.isAnimating) return;
    
    const timer = setTimeout(() => {
      dispatch({ type: 'NEXT_STEP' });
    }, state.stepDelay);
    
    return () => clearTimeout(timer);
  }, [state.isAnimating, state.currentPhase, state.currentStep]);
  
  // Reducer to handle state transitions
  function verificationReducer(state, action) {
    switch (action.type) {
      case 'START_ANIMATION':
        return { ...state, isAnimating: true };
      case 'PAUSE_ANIMATION':
        return { ...state, isAnimating: false };
      case 'NEXT_STEP':
        // Logic to determine the next step or phase
        const { nextPhase, nextStep } = calculateNextStep(state);
        return { 
          ...state, 
          currentPhase: nextPhase,
          currentStep: nextStep,
          isAnimating: nextPhase < TOTAL_PHASES
        };
      // Other actions...
    }
  }
  
  // Rest of component...
}
```

Benefits:
- Separates timing logic from state transitions
- Makes animation speed adjustable with a single value
- Easier to pause, resume, or modify the animation sequence
- Clearer progression of steps with predictable state transitions

## Alternative 2: Custom Hook with Async/Await

Another elegant approach would be to create a custom hook that manages the verification sequence:

```jsx
function useVerificationAnimation(initialDelay = 800) {
  const [phase, setPhase] = useState(0);
  const [step, setStep] = useState(0);
  const [isRunning, setIsRunning] = useState(false);
  const [delay, setDelay] = useState(initialDelay);
  
  // Define all phases and steps in a data structure
  const animationSteps = useMemo(() => [
    { name: 'DNSSEC', steps: [/* steps data */] },
    { name: 'X.509', steps: [/* steps data */] },
    { name: 'CMS', steps: [/* steps data */] }
  ], []);
  
  const totalPhases = animationSteps.length;
  const totalStepsInCurrentPhase = animationSteps[phase]?.steps.length || 0;
  
  const sleep = useCallback((ms) => new Promise(resolve => setTimeout(resolve, ms)), []);
  
  // Function to run the animation
  const runAnimation = useCallback(async () => {
    setIsRunning(true);
    
    for (let p = 0; p < animationSteps.length; p++) {
      setPhase(p);
      
      for (let s = 0; s < animationSteps[p].steps.length; s++) {
        setStep(s);
        await sleep(delay);
      }
    }
    
    setIsRunning(false);
  }, [animationSteps, delay, sleep]);
  
  return {
    phase,
    step,
    isRunning,
    delay,
    setDelay,
    runAnimation,
    currentPhaseData: animationSteps[phase],
    currentStepData: animationSteps[phase]?.steps[step],
    totalPhases,
    totalStepsInCurrentPhase
  };
}
```

Benefits:
- Clean, declarative approach that makes the animation sequence obvious
- Easily adjustable timing with a single parameter
- Animation logic is encapsulated and reusable
- Async/await pattern makes the sequence more linear and readable

Both alternatives provide a more maintainable and flexible implementation that would be easier to extend or modify in the future. The first approach is more React-idiomatic, while the second offers a cleaner sequential flow.

gnarea commented on src/components/VerificationChain/VerificationPhases.tsx:
```diff
@@ -0,0 +1,206 @@
+            />
+          </div>
+
+          <div>
+            <DirectionalArrow />
+          </div>
```
Maybe we should define the `div` inside `DirectionalArrow`? Also, all arrows are directional, so that adjective is redundant.

gnarea commented on src/components/Homepage/HowItWorks.astro:
```diff
@@ -1,4 +1,8 @@
-<section class="bg-neutral-900 px-4 sm:px-6 py-12 md:py-16 text-sm md:text-base">
+---
+import VerificationPhases from "../VerificationChain/VerificationPhases";
```
I think `VerificationChain` and `VerificationPhases` should be something like `ChainVerificationAnimation`. Also, if you rename `VerificationPhases.tsx` to `index.tsx`, you should be able to import it as `import VerificationPhases from "../ChainVerificationAnimation";`

gnarea reviewed (changes requested):
LGTM generally, except for the JS orchestrating the animation, as noted below.

Visual feedback:

- I think the verification of each step is too slow. Like 2 seconds? I think it should be 1 or so seconds.
- There's a slight lag between verification steps. I think this is due to the use of `setTimeout()`, which could be more noticeable on slower devices.
- I think it's weird that the phase column gets redrawn when the verification concludes, to add a border. Maybe the phase too should begin with an orange border?
- Similarly, it looks weird that the text at the bottom of each column moves slightly to the left when the tick mark is displayed. I suspect we may be using `display: none` instead of `display: hidden`.
- I wonder if we should have a button to restart the animation at the end. Wdyt?

maribelferreira committed:
Optimize tick icon visibility in PhaseColumn and adjust border animation duration

Refactor PhaseColumn component:
- Replace conditional rendering with CSS visibility
- Simplify SVG tick icon display logic
- Reduce border animation duration from 2s to 1s in global CSS

maribelferreira committed:
Refactor PhaseColumn and update Arrow component

Consolidate verification chain components:
- Rename PhaseColumn.tsx to Phase.tsx
- Update import references in VerificationPhases
- Remove unused description property in CMS steps
- Simplify Arrow component with reusable SVG rendering
- Restrict direction type to "down" and "right"

maribelferreira committed:
Introduce VerificationStatus enum and refactor phase rendering logic

Enhance verification chain components:
- Add VerificationStatus enum to standardize phase states
- Update PhaseColumn to use new status-based rendering
- Simplify status determination in VerificationPhases
- Modify prop types to improve type safety and readability

maribelferreira committed:
Refactor VerificationStep to use VerificationStatus enum

Update VerificationStep component:
- Replace boolean props with VerificationStatus enum
- Modify status rendering logic to use enum values
- Simplify component prop types and default values
- Improve type safety and readability of verification state management

maribelferreira committed:
Remove redundant comments in VerificationPhases and VerificationStep

Cleanup code by:
- Removing unnecessary inline comments explaining code functionality
- Preserving code structure and logic
- Improving code readability by eliminating verbose explanatory comments

maribelferreira committed:
Refactor ResponsiveArrow in VerificationPhases

Simplify Arrow component rendering:
- Wrap Arrow in a div to maintain consistent layout
- Rename DirectionalArrow to ResponsiveArrow for clarity
- Remove redundant div wrappers around ResponsiveArrow

maribelferreira committed:
Refactor verification chain components into ChainVerificationAnimation

Replace VerificationPhases with a new ChainVerificationAnimation component:
- Remove individual verification chain subcomponents
- Update HowItWorks component to use new animation component
- Simplify project structure by consolidating verification visualization logic

maribelferreira committed:
Refactor ChainVerificationAnimation with useReducer state management

Modernize animation logic by:
- Replacing useState with useReducer for complex state management
- Introducing AnimationState and AnimationAction types
- Implementing a more robust animation control mechanism
- Simplifying animation flow with a single useEffect hook
- Adding support for animation status and speed control

maribelferreira committed:
Add play/restart button to ChainVerificationAnimation

Enhance user interaction by:
- Removing pause/resume animation actions
- Adding a centered overlay button to start/restart animation
- Updating animation state management to support button interactions
- Adjusting component layout and styling for button integration

maribelferreira committed:
Update ChainVerificationAnimation with minor text refinements

Improve component details:
- Adjust CMS step title to use proper quotes
- Remove redundant comments about state types
- Simplify code by removing unnecessary inline comments

maribelferreira committed:
Add play and restart SVG icons for ChainVerificationAnimation

Enhance component with new icons:
- Import play and restart icons as raw SVG
- Replace inline SVG with dynamically rendered icons
- Add class attributes to maintain existing styling
- Simplify icon rendering using dangerouslySetInnerHTML

maribelferreira committed:
Remove animated border gradient from VerificationStep

Simplify component styling:
- Remove complex animated border gradient
- Preserve existing padding and text styling
- Reduce visual complexity of verification step rendering

maribelferreira committed:
Reintroduce animated border gradient to VerificationStep

Add dynamic border animation with:
- Conic gradient background for border
- Tailwind animation classes
- Gradient color transitions using theme colors
- Responsive border and padding styling

maribelferreira committed:
Add dark overlay and improve ChainVerificationAnimation layout

Enhance component with:
- Dark overlay for idle state
- Refined layout structure
- Improved z-index and transition handling
- Simplified flex container for verification phases

maribelferreira committed:
Update button design and layout:
- Replace text-based button with circular green button
- Separate button and text label into flex column
- Simplify play/restart icon rendering
- Adjust button sizing and styling for improved visual clarity

maribelferreira committed:
Refine ChainVerificationAnimation UI interactions and responsiveness

- Add dynamic opacity to verification phases based on animation state
- Improve play/restart button positioning and visibility
- Enhance button design with responsive sizing and placement
- Adjust overlay opacity and transition effects
- Simplify button interaction logic

maribelferreira committed:
Adjust dark overlay styling in ChainVerificationAnimation

- Increase overlay opacity from 70% to 75%
- Add rounded corners to the overlay to match component design
- Maintain existing transition and opacity state logic

maribelferreira committed:
Refactor play icon and adjust button styling in ChainVerificationAnimation

- Replace play icon SVG with a more compact, modern version
- Use raw SVG import with dynamic class injection
- Adjust dark overlay opacity to 70%
- Refine restart button positioning and sizing for responsive layout
- Simplify button and icon rendering

maribelferreira committed:
Refactor restart button rendering and padding in ChainVerificationAnimation

- Conditionally apply bottom padding based on animation state
- Render restart button only when animation is completed
- Simplify button and container conditional rendering
- Improve responsive layout and positioning of restart button

maribelferreira committed:
Refactor Phase and VerificationStep component styling

- Extract common style classes for improved readability
- Simplify border color and styling logic
- Adjust SVG tick icon sizing and margin
- Improve responsive design with consistent class naming
- Reduce code complexity by centralizing styling logic

maribelferreira committed:
Implement advanced border fill animation for verification steps

- Add custom CSS border fill animation with clip-path
- Create responsive border progress styles
- Refactor VerificationStep component to use new border fill class
- Enhance visual feedback with gradient-like border transition

maribelferreira committed:
Add rounded corners to dark overlay in ChainVerificationAnimation

- Apply rounded-md class to dark overlay div
- Maintain existing opacity and transition logic
- Ensure consistent design with component's rounded corners

gnarea commented on src/components/ChainVerificationAnimation/Phase.tsx:
```diff
@@ -0,0 +1,84 @@
+        {result}
+        <div
+          className="flex items-center"
+          dangerouslySetInnerHTML={{
+            __html: TickIcon.replace(
+              "<svg",
+              `<svg width="26" height="26" class="ml-1" style="visibility: ${
+                showTick ? "visible" : "hidden"
+              }"`,
+            ),
+          }}
```
Isn't there a "safer" alternative to `dangerouslySetInnerHTML`?

gnarea reviewed (changes requested):
Nearly there!

Suggestions:

- Step border transition looks weird: It starts to fill counter-clockwise too.
- Phase border transition isn't animated.
- Mobile/portrait: Replay button location leads to following text being pushed up and down. Maybe we should use `display: hidden`?

maribelferreira committed:
Add SVGR support for SVG imports in Astro project

- Configure vite-plugin-svgr for React SVG component imports
- Update Astro config to include SVGR plugin
- Add TypeScript declarations for SVG imports
- Refactor SVG icon rendering in ChainVerificationAnimation to use React components
- Remove manual SVG string manipulation

maribelferreira committed:
Refine border fill animation and clip-path for verification steps

- Update VerificationStep component border styling
- Modify clip-path animation in global CSS for smoother border fill effect
- Adjust animation keyframes to create more dynamic border transition
- Simplify border class application in component

maribelferreira committed:
Move border fill animation CSS to dedicated component file

- Create BorderAnimation.css for border fill animation styles
- Remove animation styles from global.css
- Import BorderAnimation.css in VerificationStep component
- Maintain existing animation logic and responsive design

gnarea reviewed (approved):
🎉
